torchgeo/datasets/geo.py

```diff
@@ -776,19 +799,97 @@ def __getitem__(self, query: GeoSlice) -> dict[str, Any]:
+                        convert_poly_coords(
+                            shapely.geometry.shape(s[0]), transform, inverse=True
+                        )
+                        for s in shapes
+                    ]
+                    px_shapes = [
+                        (shapely.clip_by_rect(p, 0, 0, width, height))
+                        for p in px_shapes
+                    ]
+
+                    # Get labels
+                    labels = np.array([s[1] for s in shapes]).astype(np.int32)
+
+                    # xmin, ymin, xmax, ymax format
+                    boxes_xyxy = np.array(
+                        [
+                            [p.bounds[0], p.bounds[1], p.bounds[2], p.bounds[3]]
+                            for p in px_shapes
+                        ]
+                    ).astype(np.float32)
+
+                case 'instance_segmentation':
+                    # Get boxes for object detection or instance segmentation
+                    px_shapes = [
+                        convert_poly_coords(
+                            shapely.geometry.shape(s[0]), transform, inverse=True
+                        )
+                        for s in shapes
+                    ]
+                    px_shapes = [
+                        (shapely.clip_by_rect(p, 0, 0, width, height))
+                        for p in px_shapes
+                    ]
+
+                    # Get labels
+                    labels = np.array([s[1] for s in shapes]).astype(np.int32)
+
+                    # xmin, ymin, xmax, ymax format
+                    boxes_xyxy = np.array(
+                        [
+                            [p.bounds[0], p.bounds[1], p.bounds[2], p.bounds[3]]
+                            for p in px_shapes
+                        ]
+                    ).astype(np.float32)
+
+                    masks = rasterio.features.rasterize(
+                        [(s[0], i + 1) for i, s in enumerate(shapes)],
+                        out_shape=(round(height), round(width)),
+                        transform=transform,
+                    )
+
+                    obj_ids = np.unique(masks)
+
+                    # first id is the background, so remove it
+                    obj_ids = obj_ids[1:]
+
+                    # convert (H, W) mask a set of binary masks
+                    masks = (masks == obj_ids[:, None, None]).astype(np.uint8)
         else:
             # If no features are found in this key, return an empty mask
             # with the default fill value and dtype used by rasterize
             masks = np.zeros((round(height), round(width)), dtype=np.uint8)
+            boxes_xyxy = np.empty((0, 4), dtype=np.float32)
+            labels = np.empty((0,), dtype=np.int32)
 
         # Use array_to_tensor since rasterize may return uint16/uint32 arrays.
-        masks = array_to_tensor(masks)
+        sample: dict[str, Tensor | CRS | GeoSlice] = {'crs': self.crs, 'bounds': query}
```

Comment:
```suggestion
        sample: dict[str, Any] = {'crs': self.crs, 'bounds': query}
```
At some point I plan to use TypedDict but not now